[PT-D][Sharding] Enable more ops needed in the transformer model training

fduwjj opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #77214

From the code base of MetaSeq Model, we have found that loads of ops are not supported by sharded tensor. In https://github.com/pytorch/pytorch/pull/75374, we have enabled most of ops already and this PR/diff aims at enabling the rest of them.

Fix some unit test errors.

Differential Revision: [D36302780](https://our.internmc.facebook.com/intern/diff/D36302780/)

Commit:
[PT-D][Sharding] Enable more ops needed in the transformer model training

From the code base of MetaSeq Model, we have found that loads of ops are not supported by sharded tensor. In https://github.com/pytorch/pytorch/pull/75374, we have enabled most of ops already and this PR/diff aims at enabling the rest of them.

Fix some unit test errors.

Differential Revision: [D36302780](https://our.internmc.facebook.com/intern/diff/D36302780/)

[ghstack-poisoned]

Commit:
Update on "[PT-D][Sharding] Enable more ops needed in the transformer model training"


From the code base of MetaSeq Model, we have found that loads of ops are not supported by sharded tensor. In https://github.com/pytorch/pytorch/pull/75374, we have enabled most of ops already and this PR/diff aims at enabling the rest of them.

Fix some unit test errors.

Differential Revision: [D36302780](https://our.internmc.facebook.com/intern/diff/D36302780/)

[ghstack-poisoned]

wanchaol commented on torch/distributed/_shard/sharded_tensor/_ops/math_ops.py:
```diff
@@ -83,21 +83,25 @@ def binary_math_op(types, args=(), kwargs=None, pg=None):
     # add
     torch.add,
     Tensor.add,
+    Tensor.add_,
```
is this working? I think we need to add a `add_` method to ShardedTensor class and dispatch to here to make it really work

fduwjj replied:
Hmm, that's wired because the unit test does pass.

wanchaol commented on test/distributed/_shard/sharded_tensor/ops/test_matrix_ops.py:
```diff
@@ -158,9 +132,6 @@ def _test_masked_fill_with_sizes(self, mask_size, broadcast_style=False):
             tensor_m = tensor.masked_fill(mask, 25.0)
             st_expected = _shard_tensor(tensor_m, spec)
-            st_expected._metadata.shards_metadata.sort(
-                key=lambda x: x.shard_offsets[0],
```
why we don't need sort this time?

fduwjj replied:
Because your PR(https://github.com/pytorch/pytorch/pull/75991) has fixed it so we now don't need it anymore.

wanchaol commented on torch/distributed/_shard/sharding_spec/chunk_sharding_spec_ops/_common.py:
```diff
@@ -57,7 +57,7 @@ def _register_sharded_op_on_local_tensor(
     @sharded_op_impl(op)
     @_sharded_op_common(op, early_stop_func, extra_check)
-    def sharded_tensor_op_on_local_shards(types, args=(), kwargs=None, pg=None):
+    def sharded_tensor_op_on_local_tensor(types, args=(), kwargs=None, pg=None):
```
hmmm we didn't change the callsites of `_register_sharded_op_on_local_shards` to `_register_sharded_op_on_local_tensor`, is this already been changes somewhere else?

fduwjj replied:
Nope, we don't need to change the callsite because this is the func inside the decorator. And we are not changing the signature of decorator itself.

wanchaol commented on test/distributed/_shard/sharded_tensor/ops/test_matrix_ops.py:
```diff
@@ -132,23 +123,6 @@ def test_sharded_tensor_softmax(self):
-    @with_comms(init_rpc=True)
-    @skip_if_lt_x_gpu(TEST_GPU_NUM)
-    @requires_nccl()
-    def test_sharded_tensor_softmax_error(self):
```
are we supporting softmax on sharding dim already, why we delete this test?

wanchaol reviewed (approved):
looks good to me! Thanks for fixing

pritamdamania87 commented on torch/distributed/_shard/sharded_tensor/_ops/matrix_ops.py:
```diff
@@ -0,0 +1,67 @@
+)
+
+
+def sharded_type_as_check(*args, **kwargs):
+    """
+    Perform extra checks for the sharded_type_as op such as the input needs to
+    be either a Tensor or ShardedTensor.
+
+    Args: same as ``torch.Tensor.type_as``.
+
+    Return: None
+    """
+    if len(args) < 2:
+        raise ValueError("Needs to give a tensor to cast type as!")
+    if not isinstance(args[1], torch.Tensor) and not isinstance(args[1], ShardedTensor):
+        raise ValueError("Needs to give a Tensor or ShardedTensor to cast type as!")
+
+
+def same_dtype(*args, **kwargs):
```
Weren't these already added in https://github.com/pytorch/pytorch/pull/75374?

fduwjj replied:
Yep. It somehow gets removed during rebase? I also felt wired about `type_as`. Originally the test fails so I have to add them back.

pritamdamania87 commented on torch/distributed/_shard/sharding_spec/chunk_sharding_spec_ops/softmax.py:
```diff
@@ -14,14 +14,14 @@ def sharded_softmax(types, args=(), kwargs=None):
         smax = torch.div(exp, exp_sum)
     else:
-        smax = torch.nn.functional.softmax(input.local_tensor())
+        smax = torch.nn.functional.softmax(input.local_tensor(), dim=dim)
```
Thanks for fixing this! Can we extend `test_softmax.py` to ensure there was a unit test covering this?

fduwjj replied:
Sure, maybe I can move this into `test_softmax.py`?

fduwjj replied:
Looks like `test_softmax.py` already contains this but it somehow use the default dim as the sharding dim so we didn't find out the corner case.

pritamdamania87 commented on torch/distributed/_shard/sharding_spec/chunk_sharding_spec_ops/matrix_ops.py:
```diff
@@ -304,3 +311,102 @@ def sharded_bmm(args, kwargs, pg):
+        if current_rank == 0
+        else None
+    )
+    st.gather(dst=0, out=global_tensor)
```
This seems pretty inefficient, can't we only do one allreduce to compute the `mean` and then given the `mean` we can compute the `standard deviation` as well and do rest of the operations just locally?

fduwjj replied:
I see. sure, let me change it like that.

Commit:
Update on "[PT-D][Sharding] Enable more ops needed in the transformer model training"


From the code base of MetaSeq Model, we have found that loads of ops are not supported by sharded tensor. In https://github.com/pytorch/pytorch/pull/75374, we have enabled most of ops already and this PR/diff aims at enabling the rest of them.

Fix some unit test errors.

Differential Revision: [D36302780](https://our.internmc.facebook.com/intern/diff/D36302780/)

[ghstack-poisoned]

fduwjj commented:
Rebase and addressed the comment from reviewers.

pritamdamania87 commented on torch/distributed/_shard/sharding_spec/chunk_sharding_spec_ops/matrix_ops.py:
```diff
@@ -304,3 +313,100 @@ def sharded_bmm(args, kwargs, pg):
+    local_variant_sq = torch.square(local_tensor - global_mean).sum(
+        norm_dims, keepdim=True
+    )
+    global_variant = torch.div(all_reduce(local_variant_sq), st_size)
```
Do we need this additional allreduce? Shouldn't the global mean be sufficient to compute the global variance?

fduwjj replied:
I believe so, the way we calculate global variance still needs information from all shards. Unless there is some fancy math which can help us avoid the extra communication. Correct me if I am wrong.

pritamdamania87 commented on test/distributed/_shard/sharded_tensor/ops/test_matrix_ops.py:
```diff
@@ -258,6 +223,62 @@ def test_sharded_tensor_view_error(self):
+        specs = _chunk_sharding_specs_list_for_test([1, 2], seed=10)
+        for spec in specs:
+            tensor = torch.rand(16, 35, 26).cuda(self.rank)
+            layer_norm = torch.nn.LayerNorm((35, 26)).cuda(self.rank)
```
Can we add a test with `elementwise_affine`?

fduwjj replied:
Sure.

fduwjj commented on test/distributed/_shard/sharded_tensor/ops/test_softmax.py:
```diff
@@ -33,11 +33,11 @@ def _test_sharded_softmax(self, dim):
         local_softmax = torch.nn.functional.softmax(local_tensor, dim)
 
-        spec = ChunkShardingSpec(dim=0, placements=[f'rank:{idx}/cuda:{idx}' for idx in range(self.world_size)])
+        spec = ChunkShardingSpec(dim=1, placements=[f'rank:{idx}/cuda:{idx}' for idx in range(self.world_size)])
```
sure!

Commit:
Update on "[PT-D][Sharding] Enable more ops needed in the transformer model training"


From the code base of MetaSeq Model, we have found that loads of ops are not supported by sharded tensor. In https://github.com/pytorch/pytorch/pull/75374, we have enabled most of ops already and this PR/diff aims at enabling the rest of them.

Fix some unit test errors.

Differential Revision: [D36302780](https://our.internmc.facebook.com/intern/diff/D36302780/)

[ghstack-poisoned]

Commit:
Update on "[PT-D][Sharding] Enable more ops needed in the transformer model training"


From the code base of MetaSeq Model, we have found that loads of ops are not supported by sharded tensor. In https://github.com/pytorch/pytorch/pull/75374, we have enabled most of ops already and this PR/diff aims at enabling the rest of them.

Fix some unit test errors.

Differential Revision: [D36302780](https://our.internmc.facebook.com/intern/diff/D36302780/)

[ghstack-poisoned]

Commit:
Update on "[PT-D][Sharding] Enable more ops needed in the transformer model training"


From the code base of MetaSeq Model, we have found that loads of ops are not supported by sharded tensor. In https://github.com/pytorch/pytorch/pull/75374, we have enabled most of ops already and this PR/diff aims at enabling the rest of them.

Fix some unit test errors.

Differential Revision: [D36302780](https://our.internmc.facebook.com/intern/diff/D36302780/)

[ghstack-poisoned]
